### ISSUE TITLE
Improvement: Add caching for nuget & msbuild incremental compilation

### ISSUE THREAD

**Issue by macsux** [paketo-buildpacks/dotnet-publish#258]
Currently, the buildpack does not make any attempts to cache and reuse assets between builds. There are two main areas of improvement:

- NuGet packages: these are currently restored each time over the wire each time a build is run, which is very wasteful. Effort should be made to setup nuget cache and reuse it across builds
- Incremental compilation: `obj` folder that is implicitly created during the build process maintains a large number of artifacts. MSBuild uses it to do incremental builds, greatly speeding up the compilation of the same project between builds.

**Comment by fg-j** [comment]
These sound like great ideas. They align with things we have done elsewhere in the Paketo project where we store build artifacts in cached layers and restore them on rebuilds. This allows the language ecosystem tooling to handle proper reuse of those artifacts on its own. An example of this can be found in the [go-build buildpack](https://github.com/paketo-buildpacks/go-build/blob/f4536d32a7467caa58890ffc44112a77b69f1330/build.go#L62-L67) which saves the go build cache in its own layer.

**Comment by ForestEckhardt** [comment]
I have taken a bit of a look into setting up a reusable Nuget cache and it seems by using `$NUGET_PACKAGES` we could quite easily set up a cache for Nuget packages could be reusable between build and speed up rebuilding. After some preliminary testing it appears to be a relatively easy slot in. I think that this definitely warrants a deeper investigation than what my short spoke allowed for. 

Resources:
https://docs.microsoft.com/en-us/nuget/reference/cli-reference/cli-ref-environment-variables
https://docs.microsoft.com/en-us/nuget/consume-packages/managing-the-global-packages-and-cache-folders

**Comment by ForestEckhardt** [comment]
I have done a little more investigation into reusing the Nuget cache and there are definitely some potential drawbacks from adding it. In some of the smaller apps that I tested against I actually saw overall build times go up because restoring and exporting the layer took more time than was saved during the build. I think overall though those apps are probably not emblematic of the larger .Net application space. The addition is relatively straight forward and I think that in the long run in is probably the right thing to do.

I also took some time to look into the `obj` folder as a means to optimize our builds. This folder is generated during publish and as indicated in the original post effectively stores a compilation cache. I found it difficult to test the efficacy of reusing `obj` folder between builds but it did appear to be relatively stable even when changing small details between builds.

I think to move forward on this I would like to split this issue into 2 separate issues so that each optimization can be tracked and benchmarked more easily. I am more than happy to write these issues up and I will like out to this issue as the originator. This should hopefully allow us to get bits and pieces of this optimization in while allowing for further discussion.
